crates/engine/tree/src/tree/root.rs

```diff
@@ -106,86 +154,483 @@ where
 
     /// Handles state updates.
     fn on_state_update(
-        _view: &reth_provider::providers::ConsistentDbView<impl Send + 'static>,
-        _input: &std::sync::Arc<reth_trie::TrieInput>,
-        _state: EvmState,
+        view: ConsistentDbView<Factory>,
+        input: Arc<TrieInput>,
+        update: EvmState,
+        state: &mut HashedPostState,
+        pending_proofs: &mut VecDeque<Receiver<Result<MultiProof, ParallelStateRootError>>>,
     ) {
-        // Default implementation of state update handling
-        // TODO: calculate hashed state update and dispatch proof gathering for it.
+        let mut hashed_state_update = HashedPostState::default();
+        for (address, account) in update {
+            if account.is_touched() {
+                let hashed_address = keccak256(address);
+
+                let destroyed = account.is_selfdestructed();
+                hashed_state_update.accounts.insert(
+                    hashed_address,
+                    if destroyed || account.is_empty() { None } else { Some(account.info.into()) },
+                );
+
+                if destroyed || !account.storage.is_empty() {
+                    let storage = account.storage.into_iter().filter_map(|(slot, value)| {
+                        value.is_changed()
+                            .then(|| (keccak256(B256::from(slot)), value.present_value))
+                    });
+                    hashed_state_update
+                        .storages
+                        .insert(hashed_address, HashedStorage::from_iter(destroyed, storage));
+                }
+            }
+        }
+
+        // Dispatch proof gathering for this state update
+        let targets = hashed_state_update
+            .accounts
+            .keys()
+            .filter(|hashed_address| {
+                !state.accounts.contains_key(*hashed_address) &&
+                    !state.storages.contains_key(*hashed_address)
+            })
+            .map(|hashed_address| (*hashed_address, HashSet::default()))
+            .chain(hashed_state_update.storages.iter().map(|(hashed_address, storage)| {
+                (*hashed_address, storage.storage.keys().copied().collect())
+            }))
+            .collect::<HashMap<_, _>>();
+
+        let (tx, rx) = mpsc::sync_channel(1);
+        rayon::spawn(move || {
+            let result = ParallelProof::new(view, input).multiproof(targets);
+            if let Err(ref e) = result {
+                error!(target: "engine::root", error = ?e, "Could not calculate multiproof");
+            }
+
+            if let Err(e) = tx.send(result) {
+                error!(target: "engine::root", error = ?e, "Could not send multiproof result");
+            }
+        });
+
+        pending_proofs.push_back(rx);
+
+        state.extend(hashed_state_update);
+    }
+
+    fn run(mut self) -> StateRootResult {
+        let mut task_state = StateRootTaskState::Idle(MultiProof::default(), B256::default());
```

Comment:
> looks like we're sending a few things around and need to await things from different event variants from different sources, hence the try_recv to not block indefinitely on one kind of receiver?

yes, exactly

> I assume we can make this work with only one channel if we unify the message variants into one single enum?

yep sounds good, will prepare changes for this

Reply:
ok done, the main changes are:
* we have now a `StateRootMessage` enum with all the message variants for state change inside the task.
* the task creates an internal channel for sending and receive messages of type `StateRootMessage`
* the task also creates a background thread that just receives items from the input state stream, wraps them into the correspondent `StateRootMessage` variant and forwards them to the internal `StateRootMessage` channel. It calls `recv` in the external channel receiver, so that it yields
* `run` now  calls `recv` in the internal `StateRootMessage` receiver (yielding) and processes each message
* there's a separate `ProofSequencer` type to manage proofs in order.

@mattsse ptal, still not tested on the mainnet machine, but local checks are looking good

Reply:
this makes sense, wondering if we could get rid of the stream message thread that just delegates messages

Reply:
for sure, done here https://github.com/paradigmxyz/reth/pull/12458/commits/3b916f69cb9af2737a8340b4392366fac34ee904